### PR TITLE
CTFE: throw unsupported error when partially overwriting a pointer

### PR DESCRIPTION
**Pull request by RalfJung**
Currently, during CTFE, when a write to memory would overwrite parts of a pointer, we make the remaining parts of that pointer "uninitialized". This is probably not what users expect, so if this ever happens they will be quite confused about why some of the data just vanishes for seemingly no good reason.
So I propose we change this to abort CTFE when that happens, to at last avoid silently doing the wrong thing.
Cc https://github.com/rust-lang/rust/issues/87184

Our CTFE test suite still seems to pass. However, we should probably crater this, and I want to do some tests with Miri as well.

### CODE REVIEW

**Comment by RalfJung** [comment]
Ah turns out Miri really needs this:
```
error: unsupported operation: unable to overwrite parts of a pointer in memory at alloc1390
   --> /home/r/src/rust/rustc/library/std/src/panicking.rs:401:13
    |
401 |             data.r = ManuallyDrop::new(f());
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unable to overwrite parts of a pointer in memory at alloc1390
    |
    = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support

    = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/r/src/rust/rustc/library/std/src/panicking.rs:401:13
    = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /home/r/src/rust/rustc/library/std/src/panicking.rs:365:19
```

**Comment by RalfJung** [comment]
@craterbot check

**Comment by oli-obk** [comment]
All of the regressions are spurious (some python network error and some file not found)

**Comment by RalfJung** [comment]
Nice, that's what I hoped for. :)

So how do we proceed here -- what are your thoughts on this PR, @oli-obk?

File: compiler/rustc_middle/src/mir/interpret/error.rs [hunk 1]
```diff
@@ -404,6 +404,9 @@ pub enum UnsupportedOpInfo {
     NoMirFor(DefId),
     /// Encountered a pointer where we needed raw bytes.
     ReadPointerAsBytes,
+    /// Overwriting parts ofa  pointer; the resulting state cannot be represented in our
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/error.rs:407]:
```suggestion
    /// Overwriting parts of a pointer; the resulting state cannot be represented in our
```

**Review by oli-obk — COMMENTED** [review]
I like how this allows us to experiment on miri first!

lgtm modulo some tests

**Comment by RalfJung** [comment]
Added a test, please have a look. :)

**Pull request merged**